bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -0,0 +1,1049 @@
+                @SuppressWarnings("null")
+                List<String> receivedLines = reader.lines().collect(Collectors.toList());
+
+                // a blank line marks the end of an SSE message
+                boolean endOfMessage = !receivedLines.isEmpty()
+                        && receivedLines.get(receivedLines.size() - 1).isBlank();
+
+                if (endOfMessage) {
+                    eventData.reset();
+                    // receipt of ANY message means the event stream is established
+                    if (!completable.isDone()) {
+                        completable.complete(Boolean.TRUE);
+                    }
+                    // append any 'data' field values to the event message
+                    StringBuilder eventContent = new StringBuilder();
+                    for (String receivedLine : receivedLines) {
+                        if (receivedLine.startsWith("data:")) {
+                            eventContent.append(receivedLine.substring(5).stripLeading());
+                        }
+                    }
+                    // note: StringBuilder.isEmpty() only available from Java 15+
+                    if (eventContent.length() > 0) {
+                        onEventData(eventContent.toString().trim());
+                    }
+                }
+            }
+            callback.succeeded();
+        }
+
+        @Override
+        public boolean onIdleTimeout(@Nullable Stream stream, @Nullable Throwable x) {
+            return false;
+        }
+
+        @Override
+        public void onReset(@Nullable Stream stream, @Nullable ResetFrame frame) {
+            handleHttp2Error(Http2Error.RESET);
+        }
+    }
+
+    /**
+     * Enum of potential fatal HTTP 2 session/stream errors.
+     */
+    private enum Http2Error {
+        CLOSED,
+        FAILURE,
+        TIMEOUT,
+        RESET,
+        IDLE,
+        GO_AWAY,
+        UNAUTHORIZED;
+    }
+
+    /**
+     * Private exception for handling HTTP 2 stream and session errors.
+     */
+    @SuppressWarnings("serial")
+    private static class Http2Exception extends ApiException {
+        public final Http2Error error;
+
+        public Http2Exception(Http2Error error) {
+            this(error, null);
+        }
+
+        public Http2Exception(Http2Error error, @Nullable Throwable cause) {
+            super("HTTP 2 stream " + error.toString().toLowerCase(), cause);
+            this.error = error;
+        }
+    }
+
+    /**
+     * Adapter for listening to HTTP 2 session status events.
+     *
+     * The session must be permanently connected, so it ignores onIdleTimeout() events.
+     * It also handles the following fatal events by notifying the containing class:
+     *
+     * <li>onClose()</li>
+     * <li>onFailure()</li>
+     * <li>onGoAway()</li>
+     * <li>onReset()</li>
+     */
+    private class SessionListenerAdapter extends Session.Listener.Adapter {
```

Comment:
Since this file is rather large (> 1000 lines) consider extracting this class to another file.

Reply:
Umm. With @maniac103 we changed these adapter classes from being static embedded classes to being instance embedded classes so that they can directly call back the methods of the outer class. So moving the embedded classes into separate external classes would complicate the code and reduce the performance. I am not sure if code ‘prettification’ should take precedence over performance. @maniac103 WDYT?

Reply:
I admit just adding the comment for consideration without thoroughly reading the contents to understand the dependencies. We should not sacrifice performance. I will have another look to see if I could give a more detailed proposal.

Reply:
I don't think performance is an issue; some sort of callback mechanism would probably work equally well from a performance POV. IMO, the problem is rather that there's so many (logic) dependencies between the HTTP/2 adapter classes and the rest of this contents of this class that I don't yet see how splitting them out could be done in a beneficial way.
The only change I'd do if I had written this code myself would be putting the inner classes to the end of the file.

Reply:
> if I had written this code myself would be putting the inner classes to the end of the file.

@maniac103 FWIW I always use the Eclipse IDE Sort Members command on my code to keep my classes looking tidy. It sorts the various members by default in the order shown below.. :)

![image](https://user-images.githubusercontent.com/893994/233845530-8609c308-8959-43a8-97f6-5773f28d0967.png)